x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -30,134 +31,193 @@ module LogStash module Filters module Geoip class DatabaseManager
+    cc_city_database_path = get_db_path(CITY, CC)
+    cc_asn_database_path = get_db_path(ASN, CC)
+
+    @metadata = DatabaseMetadata.new
+    unless @metadata.exist?
+      @metadata.save_metadata(CITY, CC, false)
+      @metadata.save_metadata(ASN, CC, false)
     end
+
+    city_database_path = @metadata.database_path(CITY) || cc_city_database_path
+    asn_database_path = @metadata.database_path(ASN) || cc_asn_database_path
+
+    # reset md5 to allow re-download when the file is gone
+    DB_TYPES.map { |type| @metadata.reset_md5(type) if @metadata.database_path(type).nil? }
+
+    @states = { "#{CITY}" => DatabaseState.new(@metadata.is_eula(CITY),
+                                               Concurrent::Array.new,
+                                               city_database_path,
+                                               cc_city_database_path),
+                "#{ASN}" => DatabaseState.new(@metadata.is_eula(ASN),
+                                              Concurrent::Array.new,
+                                              asn_database_path,
+                                              cc_asn_database_path) }
+
+    @download_manager = DownloadManager.new(@metadata)
+
+    @trigger_download = Concurrent::AtomicBoolean.new(false)
   end
 
+  protected
+  # notice plugins to update database path to the new download
+  # update timestamp when download is valid or there is no update
+  # do daily check and clean up
   def execute_download_job
     begin
-      has_update, new_database_path = @download_manager.fetch_database
-      @database_path = new_database_path if has_update
-      @metadata.save_timestamp(@database_path)
-      has_update
+      updated_db = @download_manager.fetch_database
+      updated_db.each do |database_type, valid_download, dirname, new_database_path|
+        if valid_download
+          @metadata.save_metadata(database_type, dirname, true)
+          @states[database_type].is_eula = true
+          @states[database_type].database_path = new_database_path
+          @states[database_type].plugins.dup.each { |plugin| plugin.update_database(new_database_path) if plugin }
+        end
+      end
+
+      updated_type = updated_db.map { |database_type, valid_download, new_database_path| database_type }
+      (DB_TYPES - updated_type).each { |unchange_type| @metadata.update_timestamp(unchange_type) }
     rescue => e
       logger.error(e.message, :cause => e.cause, :backtrace => e.backtrace)
+    ensure
       check_age
-      false
+      clean_up_database
+    end
+  end
+
+  # call expiry action if database is expired and EULA
+  def check_age(database_types = DB_TYPES)
+    database_types.map do |database_type|
+      days_without_update = (::Date.today - ::Time.at(@metadata.updated_at(database_type)).to_date).to_i
+
+      case
+      when days_without_update >= 30
+        if @states[database_type].is_eula && @states[database_type].plugins.size > 0
```

Comment:
(5) this check ensures Logstash can use CC database indefinitely. we only notify plugins to do expiry action when the plugin is using EULA and age >= 30